app/test_main.py

```diff
@@ -1 +1,31 @@
-# write your code here
+import datetime
+from unittest import mock
+
+import pytest
+
+from app.main import outdated_products
+
+
+@pytest.fixture()
+def products_template() -> list:
+    return [
+        {
+            "name": "salmon",
+            "expiration_date": datetime.date(2022, 2, 10),
+            "price": 600
+        },
+        {
+            "name": "chicken",
+            "expiration_date": datetime.date(2022, 2, 5),
+            "price": 120
+        },
+        {
+            "name": "duck",
+            "expiration_date": datetime.date(2022, 2, 1),
+            "price": 160
+        }]
+
+
+@mock.patch("app.main.datetime.date.today", return_value=datetime.date(2023, 2, 3))
+def test_check_if_all_expired(mocked_time, products_template) -> None:
+    assert outdated_products(products_template) == ["salmon", "chicken", "duck"]
```

Comment:
The reason why this patch does not work here is because `today` is called from `datetime.date`, which is already immutable. So by the time of call, it is supposed that `datetime.date` is already created (and it is immutable) and we are calling `today` from this immutable type, trying to change the value. In case we mock `app.main.datetime` itself, we do not rely on the previously immutable `datetime.date` object in the patch path and define the return value of the method `today()` directly during the allocation of the `datetime.date`, and this change applies to the function scope as well.